File: cpp/_103.cpp
// O(N) / O(N)
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) return {};
        
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);
        bool left_to_right = true;
        
        while (!q.empty()) {
            int level_size = q.size();
            vector<int> cur_level(level_size, 0);
            for (int i = 0; i < level_size; i++) {
                TreeNode* node = q.front();
                q.pop();
                int idx = left_to_right ? i : level_size - i - 1;
                cur_level[idx] = node->val;
                if (node->left)
                    q.push(node->left);
                if (node->right)
                    q.push(node->right);
            }
            left_to_right = !left_to_right;
            ans.push_back(cur_level);
        }
        
        return ans;
    }
};

File: python/_103.py
# O(N) / O(N)
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []

        ans = []
        queue = collections.deque([root])
        left_to_right = True
        
        while queue:
            level_size = len(queue)
            cur_level = [0] * level_size
            for i in range(level_size):
                node = queue.popleft()
                idx = i if left_to_right else level_size - i - 1
                cur_level[idx] = node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            left_to_right = not left_to_right
            ans.append(cur_level)
        return ans